### PR TITLE
Fix parse obj root

### PR DESCRIPTION
**Pull request by dmontagu**
## Change Summary

Makes nested root models work properly. In particular, the code raising errors in #1190 doesn't raise errors with this patch. I suspect this may resolve other issues related to custom root models as well.

Also required a new flag to make nested root models generate proper json.

I believe all of the newly added checks for the `__custom_root_type__` could be moved to model creation time by making use of some method-wrapping logic in the metaclass.

The fact that the `unpack_custom_root` argument is passed in the various recursive calls to `.dict` is the only performance thing that I think is likely hard to eliminate (without breaking the json generation). But it *would* be possible to move the actual *check* (and subsequent unwrapping) into a wrapper function so that it is only performed for types that actually have a custom root. I'm not sure if adding another argument itself adds an amount of performance overhead that is worth fretting over. (But I would also understand if you don't want to add the flag at all.)

## Related issue number

Addresses #1190 

## Checklist

* [ ] Unit tests for the changes exist
* [ ] Tests pass on CI and coverage remains at 100%
* [ ] Documentation reflects the changes where applicable
* [ ] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)

### CODE REVIEW

File: tests/test_parse.py [hunk 1]
```diff
@@ -151,3 +151,25 @@ class Model(BaseModel):
...
+    obj = NestedRootMapping.parse_obj({'1': {'2': {}}})
+    assert obj.dict() == {'__root__': {1: {'__root__': {2: {'__root__': {}}}}}}
+    assert obj.json() == '{"1": {"2": {}}}'
+
+
+def test_parse_nested_custom_root_tuple():
+    class NestedRootTuple(BaseModel):
+        __root__: Tuple[int, Optional["NestedRootTuple"]]
+
+    NestedRootTuple.update_forward_refs()
```
Comment (dmontagu) [inline tests/test_parse.py]: I discovered a strange bug here -- if you don't call this `update_forward_refs`, it doesn't raise an error, but then below you actually get:

```python
    obj = NestedRootTuple.parse_obj(('1', ('2', ('3', None))))
    assert obj.dict() == {'__root__': (1, {'__root__': ('2', {'__root__': ('3', None)})})}  # note 2 and 3 became strings
```

Might be worth looking into separately.

Comment (dmontagu) [reply tests/test_parse.py]: Yep, I confirmed this is not related to custom root; creating a separate issue.

**Comment by samuelcolvin** [comment]
This seems like a backwards incompatible change to `__init__` logic? So I guess, if we make this change, we should wait for v2.

I'm also concerned about the madness of `dict(unpack_custom_root=True)` not returning a dict, but then again I'm conscious that `dict()` and `json()` should have the same (or as similar as possible) interfaces and `json()` need to return the root object not `{"__root__": thing}` in v2.

In summary, are we sure there isn't a similar way of fixing the original problem?

**Comment by dmontagu** [comment]
I think the main problem is that there are some places in the code where object initialization happens via `Model(**data)` or similar. This generally makes it difficult to handle issues with a custom root model properly because validation *requires* a dict.

I'm not saying that a backwards-compatible fix isn't possible, but I think it would help a lot if there was a single model-validation logic path that involved passing a raw object, rather than making use of the `__init__` function and requiring a dict as input. It seems like some of these issues might be easier to address, for example, if parse_obj was the main way to create a validated instance, and `__init__` delegated to that function, rather than the other way around (as it is now).

But I haven't thought through things like performance implications etc. where the current approach may have marginal benefits for the most common use cases.

At any rate, I agree that this PR is likely unworkable in the near term, so I'm going to close it.
